backends/dpdk/dpdkAsmOpt.h

```diff
@@ -290,6 +291,246 @@ class ShortenTokenLength : public Transform {
     }
 };
 
+/// This pass collect use def info by analysing all possible
+/// source and destinations, this info will be used by copy elimination pass
+class CollectUseDefInfo : public Inspector {
+    P4::TypeMap* typeMap;
+
+ public:
+    // Member expression as string considered key for all below maps, and value contains
+    // number of occurences either at use or def point.
+    // And this point it's assumed all keys are unique.
+    std::unordered_map<cstring /*member expresion as string */, int> usesInfo;
+    std::unordered_map<cstring, int> defInfo;
+    std::unordered_map<cstring /*def*/, const IR::Expression* /*use*/> replacementMap;
+    std::set<cstring> dontEliminate;
+
+    explicit CollectUseDefInfo(P4::TypeMap* typeMap) : typeMap(typeMap) {
+        dontEliminate.insert("m.pna_main_output_metadata_output_port");
+        dontEliminate.insert("m.psa_ingress_output_metadata_drop");
+        dontEliminate.insert("m.psa_ingress_output_metadata_egress_port");
+    }
+
+    bool preorder(const IR::DpdkJmpCondStatement *b) override {
+        usesInfo[b->src1->toString()]++;
+        usesInfo[b->src2->toString()]++;
+        return false;
+    }
+
+    bool preorder(const IR::DpdkLearnStatement *b) override {
+        usesInfo[b->timeout->toString()]++;
```

Comment:
Makes me wonder whether you could define some methods for the base statement class which returns a list of source and destination expressions. Then you need only one function instead of one preorder for each statement.

Reply:
we tried use base class as much possible, where source and destination are similar i.e. DpdkBinaryStatement DpdkJmpCondStatement etc..